feat(rust,python): initial working version of Decimal Series

aldanor opened the pull request.

aldanor commented:
@plaflamme @ritchie46 Pushing my WIP branch in its current state, this obviously needs further work before we even get to arithmetics, ops and the such. But first I think we should make decisions on whether things are generally correct type-wise/dtype-wise etc given that decimals are a bit weird compared to other types.

// there's also a standing bug with creating decimal series from anyvalues – just inferring the type from the first decimal doesn't work, you have to scan the entire sequence first to determine the scale

aldanor commented:
> > // there's also a standing bug with creating decimal series from anyvalues – just inferring the type from the first decimal doesn't work, you have to scan the entire sequence first to determine the scale
> 
> Maybe we could start with negative scale (indicating that it should be set) and then we keep track of it while we are scanning the items. Another strategy is always using a default scale if not set by users. We do the same with datetime units.

I think it's a bit different here from datetime units, at least if we restrict it to `decimal.Decimal` (and potentially arrow scalar `decimal128`). Those always have defined scale, e.g. you can have input of
```python
[
    Decimal('1.23'),    # scale = 2
    Decimal('-1000'),   # scale = 0
    Decimal('0.0001'),  # scale = 4
    Decimal('0'),       # scale = 0
]
```
In this case you have to have to use scale = 4, ending up with the following values:
```python
[1_2300, -1000_0000, 1, 0]
```
---
One exception would be if we allow to mix integers and Decimals in any-values, e.g. ideally this should produce the same output:
```python
[Decimal('1.23'), -1000, Decimal('0.0001'), Decimal('0') ]
```

ritchie46 commented:
@aldanor shall we try to get an absolute minimal version working and merge that in? I rather break this down in smaller steps and see what we encounter.

aldanor commented:
> @aldanor shall we try to get an absolute minimal version working and merge that in? I rather break this down in smaller steps and see what we encounter.

I think after writing at least *some* tests for the basic functionality that has been added here, like formatting, parsing converting to/from python decimal, and from arrow, to/from anyvalues etc – I would be pretty uncomfortable merging it otherwise. I can jump on adding some tests here – if you have any particular suggestions for test cases please feel free to post.

// Btw: failing Python tests on the CI are actually a sign that things sort of work as intended – they are the ones that checked that Decimal was converted to Float64 which it isn't the case anymore.

ritchie46 commented:
> I think after writing at least some tests for the basic functionality that has been added here, like formatting, parsing converting to/from python decimal, and from arrow, to/from anyvalues etc – I would be pretty uncomfortable merging it otherwise. I can jump on adding some tests here – if you have any particular suggestions for test cases please feel free to post.

The python tests are preferred test locations. These are the ultimate integation tests and don't hurt compile times. If we want to test on the rust side, the integrations tests folder: `tests/it/` is preferred. We had very slow CI builds with inline tests, that's why we move them to `tests/it`.

aldanor commented:
So, I guess conversion logic must go into `Series::from_any_values_and_dtype()`. What's a bit weird is that's it unclear what the `dtype` that's being provided is and where it comes from (because for decimals, to come up with the dtype, you need to scan the series, so it makes sense to do it in Rust).

The `dtype` in `Series::from_any_values_and_dtype()` – must it be respected strictly? I.e. if it tells you dtype==d1, can I return dtype==d2? Or try hard to fit it into d2 and fail if it can't?

Related question: where does this `dtype` even come from, is it mainly internal inference or external `dtype=...`?

I was thinking the logic should be something like this:
- Scan through anyvalues list once, compute min scale and max scale (where integers get treated as scale=0)
- If `min_scale == max_scale == dtype.scale`, all good, no conversions needed. Do another pass and convert all raw values to i128 and build Series. The last step, what if prec != None? Do we check values here to see if they fit within precision range?
- If `min_scale != max_scale`, some elements need to be rescaled. Same if either `min_scale` or `max_scale` != `dtype.scale`?... Create a new vector a rescale everything to `max_scale` if possible, then go back to the previous step.
- If some anyvalue objects, hypothetically, have prec != None, what do we do? What if some have prec = None and some have != None? How does it play with `dtype.prec`?
- I was thinking, should `AnyValue` even have `precision`? Python integers and decimals don't have precision. The only way to get it would be either from arrow scalars (?) or by indexing pl.Series itself (but if you do it in Python, you get back Python decimal and then precision is lost again, so is there a point?).

plaflamme commented on py-polars/src/conversion.rs:
```diff
@@ -500,13 +537,61 @@ impl ToPyObject for Wrap<&DateChunked> {
+        let utils = UTILS.as_ref(py);
+        let convert = utils.getattr("_to_python_decimal").unwrap();
+        let py_scale = self.0.scale().to_object(py);
+        // if we don't know precision, the only safe bet is to set it to 39
```
FMI: why is it `39` here and `38` elsewhere?

aldanor replied:
Because, technically, Python's decimals are not restricted by prec=38 (unlike arrow decimals), and... technically we could have a number between `10.pow(38)` and `i128::MAX`? (we still need to decide whether we want to allow values in that range).

plaflamme replied:
Ah, I hadn't realized python Decimals didn't have this upper bound. My opinion would be to not support more than 38. Seems like it can only introduce edge cases and the value seems pretty low?

aldanor replied:
Let's mark this as "to figure out". I'd lean towards allowing the whole i128 scale if prec=None, but you'll have to check it "at the edges" (only arrow edges though! since Python doesn't care).

plaflamme commented on py-polars/src/conversion.rs:
```diff
@@ -610,6 +694,24 @@ impl<'s> FromPyObject<'s> for Wrap<AnyValue<'s>> {
                     Ok(Wrap(AnyValue::Time(v)))
                 }),
+                "Decimal" => {
+                    // note: there seems to be no way to extract precision from decimal.Decimal
```
Correct, though I believe it's available using `getcontext()` which is a thread local. I'm not sure it's a good idea to call `getcontext()` here, but I believe you can extract precision like this (more or less):

```rust
let py_decimal = PyModule::import(py, "decimal")?;
let ctx = py_decimal.call_method("getcontext", (), None)?;
let precision = ctx.getattr("prec");
```

aldanor replied:
Nope, IIUC `getcontext()` returns you the current context used to build *_new_* decimals. It doesn't tell you anything about existing ones.

```python
print(decimal.Context(prec=2).create_decimal('1.23456'))
print(decimal.getcontext().prec)
```
```
1.2
28
```

plaflamme replied:
I see. I suppose that makes sense since it's only relevant when computing over decimals; in which case, you need `Context` for creating new decimals and rounding and such. Even more argument towards dropping `precision`!

That said, if we want *some* value, using the thread's `Context` might still be something to consider: it is the "most likely" value of `precision` and the behaviour wouldn't be surprising unless you're doing something very specific.

aldanor replied:
See some of my thoughts above, tbh I'd be inclined towards completely dropping `precision` from `AnyValue` as it's basically unused now and potentially complicates "from any-values" procedure by an order of magnitude if it's to be respected.

plaflamme reviewed (commented):
My 2 cents. This looks great BTW 👍

aldanor commented on polars/polars-core/src/datatypes/dtype.rs:
```diff
@@ -256,8 +257,13 @@ impl Display for DataType {
+            DataType::Decimal(prec, scale) => {
+                return f.write_str(&match prec {
+                    Some(prec) => format!("decimal[{}, {}]", prec, scale),
+                    None => format!("decimal[*, {}]", scale),
```
I've thought about this too, and I don't really enjoy the look of it either :)
- scale is _way_ more important
- precision will often be just None
- if scale and precision were flipped, `precision=None` could become an optional kwarg from decimal datatype

So we could do...
- Something like `decimal[5]` (scale=5, prec=None) and `decimal[5(2)]` (scale=5, prec=2)
- But the above probably only makes sense if we swap prec/scale everywhere (DataType in Rust, Decimal datatype in Python, etc), otherwise it's even more confusing....

aldanor replied:
Another middle-ground solution would be not flipping prec/scale, but printing it as `decimal[5]` if prec is None or something like `decimal{2}[5]` if prec=2. Idk, still a bit ugly and possibly confusing 🤔

plaflamme replied:
`decimal[5(2)]` LGTM as well.

An alternative for showing `scale`, `precision` in that order would be to use a dot in front of scale making it more explicit, e.g.: `decimal[.5,2]` Then various separators could be used, e.g.: `:` or `;` perhaps.

> But the above probably only makes sense if we swap precision/scale everywhere (DataType in Rust, Decimal datatype in Python, etc), otherwise it's even more confusing....

Yeah, good point, we'd have to flip it around everywhere. Probably not worth it. It would also mean it'd be reversed from Arrow's definition which can be somewhat confusing.

No strong opinion. Using `_` would probably pass as "pythonic" as well.

aldanor replied:
@plaflamme Ok, changed it to `decimal[2]` and `decimal[.5,2]`.

aldanor commented:
@ritchie46 @plaflamme Here's a new batch of updates (the whole thing is brittle as f, I'm surprised it actually works but it does; lots of work ahead to cover various corners!)

- `DataType::Decimal` now has an optional scale (heh, we've come full circle). It's a bit weird because for instantiated Series (or DecimalChunked) it will _never_ be None, however it is required to pass the information downstream that "Decimal scale has to be inferred". I initially used some placeholders like usize::MAX, but I'd prefer for it to crash loudly if something wrong happens.
- Removed `precision` from `AnyValue::Decimal`.
- All existing tests now pass (had to add special support for dataclasses).
- Added a few very basic tests (instantiation of series/frames with a mixture of decimals, ints and nones, and then recovering them back into python)
- Supertype merging logic and constructor type fallback logic is now implemented. This means an array of decimals + ints + nones will be casted to decimals, regardless of order.
- Reworked `&[AnyValue]` -> Decimal logic completely, with a double-pass approach (no allocations/ops if no conversions needed though).
- Conversion to Decimal series will now do a precision check (still not sure if we should check for 38 digits if precision=None).
- Some please take a detailed look at `any_values_to_decimal()` conversion routine, that one wasn't so easy to figure out...?

aldanor commented:
Next steps, I think: write tests for all possible ways to build series/frames with decimals in them from Python/Arrow, and to convert out of it. There are definitely gaps to fill and decisions to make (e.g., should pl.Series(['0.01'], dtype=pl.Decimal) yield decimal series? would be cool if it did. What's the supertype of Decimal and str? not sure).

ritchie46 commented:
> DataType::Decimal now has an optional scale (heh, we've come full circle). It's a bit weird because for instantiated Series (or DecimalChunked) it will never be None, however it is required to pass the information downstream that "Decimal scale has to be inferred". I initially used some placeholders like usize::MAX, but I'd prefer for it to crash loudly if something wrong happens.

This is also the case for `Categorical`. The data types can hold information for that column that is only known after the column is created. :+1: 

> Removed precision from AnyValue::Decimal.

Yes, that should be in `DataType`. `AnyValue` should only have varying data. The constant data per column should be in `DataType`.

ritchie46 commented on py-polars/src/conversion.rs:
```diff
@@ -178,6 +178,21 @@ fn struct_dict<'a>(
 }
 
+fn decimal_to_digits(v: i128, buf: &mut [u8]) -> usize {
+    const ZEROS: i128 = 0x3030_3030_3030_3030_3030_3030_3030_3030;
```
Can you add some comments explaining what this code does and why the `unsafe` is sound?

aldanor replied:
Actually you're right, `buf` should be `&mut [u8; 48]` here, fixed (then unsafe is obviously sound; we could use something like `bytemuck` here I guess to hide unsafe but it's sort of the same thing).

I'll add a few basic tests for this on the Python side as well.

aldanor commented on py-polars/polars/utils.py:
```diff
@@ -136,6 +137,17 @@ def _timedelta_to_pl_timedelta(td: timedelta, tu: TimeUnit | None = None) -> int
         raise ValueError(f"tu must be one of {{'ns', 'us', 'ms'}}, got {tu}")
 
 
+_create_decimal = [Decimal] + [Context(prec=p).create_decimal for p in range(1, 63)]
```
Micro-optimization habit :) It's 1.5x faster this way.

But you're right re: lean imports, we can just `lru_cache` it, it then seems to be about the same perf-wise.

aldanor commented on polars/polars-core/src/chunked_array/logical/decimal.rs:
```diff
@@ -4,34 +4,91 @@ use crate::prelude::*;
 
+    #[inline]
+    fn get_any_value(&self, i: usize) -> PolarsResult<AnyValue<'_>> {
+        self.0
```
Makes sense, fixed (and removed `AV::into_decimal`, not needed anymore).

aldanor commented on polars/polars-core/src/chunked_array/logical/decimal.rs:
```diff
@@ -4,34 +4,91 @@ use crate::prelude::*;
+
+    #[inline]
+    unsafe fn get_any_value_unchecked(&self, i: usize) -> AnyValue<'_> {
+        self.0.get_any_value_unchecked(i).into_decimal(self.scale())
```
Fixed.

ritchie46 commented on py-polars/src/conversion.rs:
```diff
@@ -610,6 +694,24 @@ impl<'s> FromPyObject<'s> for Wrap<AnyValue<'s>> {
+                    let (sign, digits, exp): (i8, Vec<u8>, i32) =
+                        ob.call_method0("as_tuple").unwrap().extract().unwrap();
+                    // (using Vec<u8> is not the most efficient thing here... the input is a tuple)
+                    // let digits = digits.into_iter().map(|d| d.extract::<u8>().unwrap());
```
I believe this is a left-over.

ritchie46 commented on py-polars/src/conversion.rs:
```diff
@@ -610,6 +694,24 @@ impl<'s> FromPyObject<'s> for Wrap<AnyValue<'s>> {
+                        ob.call_method0("as_tuple").unwrap().extract().unwrap();
+                    // (using Vec<u8> is not the most efficient thing here... the input is a tuple)
+                    // let digits = digits.into_iter().map(|d| d.extract::<u8>().unwrap());
+                    // let digits = digits.iter().map(|s| *s as u8).collect::<Vec<_>>();
```
I believe this is a left-over.

ritchie46 commented on py-polars/polars/datatypes.py:
```diff
@@ -281,6 +281,32 @@ def __repr__(self) -> str:
 
 
+class Decimal(DataType):
+    """Decimal 128-bit type with an optional precision and non-negative scale."""
```
Can we explicitly state that the `Decimal` type is in `alpha` state.

ritchie46 reviewed (commented):
Alright. Thanks a lot @aldanor! This is great functionality to have. I have left a few minor remarks and then it is good to go.

ritchie46 commented:
Thanks a lot @aldanor. Is it ok if I ping you in future decimal related issues?

aldanor commented:
> Thanks a lot @aldanor. Is it ok if I ping you in future decimal related issues?

@ritchie46 👍

marsupialtail commented:
Some of my workloads are now breaking because polars infer things to be decimal type and then proceeds to complain that some operations are not supported on the decimal column. This makes me unable to use latest 0.16.10, everything works for me 0.16.8.FYI @ritchie46 @aldanor

ritchie46 commented:
Can you tell which operations are missing?
